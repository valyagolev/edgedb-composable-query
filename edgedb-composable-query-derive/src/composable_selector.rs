use crate::{opts::ComposableQueryOpts, query::QueryVar, tokens::ComposableQueryAttribute};
use darling::FromDeriveInput;
use quote::quote;
use syn::DeriveInput;

pub fn derive_composable_selector_impl(
    item: DeriveInput,
    selector_only: bool,
) -> darling::Result<proc_macro2::TokenStream> {
    let item = ComposableQueryOpts::from_derive_input(&item)?;
    let mut attribs = ComposableQueryAttribute::from_attrs(&item.attrs)?;

    if selector_only {
        attribs.push(ComposableQueryAttribute::Select(QueryVar::Var(
            "".to_string(),
        )));
    }

    let query = ComposableQueryAttribute::into_query(attribs, &item.data)?;
    let selector = &query.result;
    let ident = &item.ident;

    let result_type = selector.as_composable_query_result_type();

    Ok(quote! {
        impl ::edgedb_composable_query::composable::EdgedbComposableSelector for #ident {
            const RESULT_TYPE: ::edgedb_composable_query::composable::ComposableQueryResultKind =
                #result_type;

            fn format_selector(fmt: &mut impl ::std::fmt::Write) -> Result<(), std::fmt::Error> {
                use ::edgedb_composable_query::__itertools::Itertools;

                #selector

                Ok(())
            }
        }
    })
}

#[cfg(test)]
fn derive_composable_selector_for_test(
    item: proc_macro2::TokenStream,
) -> darling::Result<proc_macro2::TokenStream> {
    use syn::DeriveInput;

    let item = syn::parse2::<DeriveInput>(item)?;

    derive_composable_selector_impl(item, true)
}
