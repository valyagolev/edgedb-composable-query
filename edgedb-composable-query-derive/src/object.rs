use darling::ast::{self};

use darling::{util, FromDeriveInput, FromField};

use quote::quote;

use syn::{DeriveInput, Type};

#[derive(Debug, FromField)]
// #[darling(attributes(lorem))]
pub struct EdgedbObjectField {
    ident: Option<syn::Ident>,
    ty: Type,
    // #[darling(default)]
    // skip: bool,
}

#[derive(Debug, FromDeriveInput)]
#[darling(forward_attrs(allow, doc, cfg))]
struct EdgedbObjectOpts {
    ident: syn::Ident,
    attrs: Vec<syn::Attribute>,
    data: ast::Data<util::Ignored, EdgedbObjectField>,
}

pub fn derive_edgedb_object_impl(item: DeriveInput) -> darling::Result<proc_macro2::TokenStream> {
    let item = EdgedbObjectOpts::from_derive_input(&item)?;
    let fields = item.data.take_struct().ok_or_else(|| {
        darling::Error::custom("expected struct with named fields").with_span(&item.ident)
    })?;

    let field_names = fields
        .iter()
        .map(|f| f.ident.as_ref().unwrap())
        .collect::<Vec<_>>();

    let item_name = &item.ident;

    Ok(quote! {

        impl EdgedbObject for #item_name {

            fn from_edgedb_object(
                shape: edgedb_protocol::codec::ObjectShape,
                mut fields: Vec<Option<edgedb_protocol::value::Value>>,
            ) -> anyhow::Result<Self> {
                use edgedb_composable_query::EdgedbSetValue;

                #(
                    let mut #field_names = None;
                )*;

                for (i, s) in shape.elements.iter().enumerate() {
                    match s.name.as_str() {
                        #(
                            stringify!(#field_names) => {
                                #field_names = fields[i]
                                    .take()
                                    .map(EdgedbSetValue::from_edgedb_set_value)
                                    .transpose()?;
                            }
                        )*
                        _ => {}
                    }
                }

                Ok(Self {
                    #(
                        #field_names: EdgedbSetValue::interpret_possibly_missing_required_value(#field_names)?,
                    )*
                })
            }
        }
    })
}

#[cfg(test)]
fn derive_edgedb_object_for_test(
    item: proc_macro2::TokenStream,
) -> darling::Result<proc_macro2::TokenStream> {
    let item = syn::parse2::<DeriveInput>(item)?;

    derive_edgedb_object_impl(item)
}

#[cfg(test)]
mod test {
    use proc_macro2::TokenStream;
    use quote::quote;

    use super::derive_edgedb_object_for_test;

    fn on_one_quote(input: TokenStream) -> String {
        let out = derive_edgedb_object_for_test(input).unwrap();

        let s = out.to_string();
        let as_file = match syn::parse_file(&s) {
            Ok(f) => f,
            Err(e) => {
                println!("{}", s);
                panic!("failed to parse output: {}", e);
            }
        };

        prettyplease::unparse(&as_file)
    }

    #[test]
    fn insta_test_tuple_direct() {
        let input = quote! {

            #[derive(Debug, PartialEq, EdgedbObject)]
            struct ExamplImplStruct {
                a: String,
                b: Option<String>,
            }

        };

        let formatted = on_one_quote(input);

        insta::assert_snapshot!(formatted);
    }
}
