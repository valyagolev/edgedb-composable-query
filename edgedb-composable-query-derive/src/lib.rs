use composable_selector::derive_composable_selector_impl;
use object::derive_edgedb_object_impl;
use syn::DeriveInput;

mod composable_query;
mod composable_selector;
mod object;
mod opts;
mod query;
mod selector;
mod tokens;

#[proc_macro_derive(EdgedbObject)]
pub fn derive_edgedb_object(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let item = syn::parse_macro_input!(item as DeriveInput);

    match derive_edgedb_object_impl(item) {
        Ok(ts) => ts.into(),
        Err(e) => e.write_errors().into(),
    }
}

#[proc_macro_derive(EdgedbComposableSelector, attributes(params, with, var))]
pub fn derive_edgedb_composable_selector(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let item = syn::parse_macro_input!(item as DeriveInput);

    match derive_composable_selector_impl(item, true) {
        Ok(ts) => ts.into(),
        Err(e) => e.write_errors().into(),
    }
}
