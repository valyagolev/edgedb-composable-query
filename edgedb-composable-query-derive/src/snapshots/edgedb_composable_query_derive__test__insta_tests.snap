---
source: edgedb-composable-query-derive/src/lib.rs
assertion_line: 148
expression: formatted
---
fn query_gen(mut arg_n: usize, suffix: &str) -> String {
    let withs = vec![];
    {
        withs.push((n, format!("${arg_n}")));
    }
    {
        withs.push((v, format!("${arg_n}")));
    }
}
fn my_query(n: i32, v: String) -> (String, InsertQ) {
    todo!()
}

